### ISSUE TITLE
Incorrect error reporting when a window function is used as a scalar in WHERE clause: ClassCastException SqlWindowFunction cannot be cast to class SqlScalarFunction

### ISSUE THREAD

**Issue by miniway** [trinodb/trino#6447]
We expect a syntax error when window function is used at filter condition but it raises INTERNAL_ERROR
```
presto:sf1> select row_number() over() from information_schema.tables where row_number() > 5 ;
Query 20201218_082030_00005_64g2b failed: java.lang.ClassCastException: class io.prestosql.operator.window.SqlWindowFunction cannot be cast to class io.prestosql.metadata.SqlScalarFunction (io.prestosql.operator.window.SqlWindowFunction and io.prestosql.metadata.SqlScalarFunction are in unnamed module of loader 'app')
```

An old version, 317, shows `row_number():bigint is not a scalar function`, but it was INTERNAL_ERROR also

**Comment by findepi** [comment]
full stacktrace (based on current master)

```
presto:default> select row_number() over() from information_schema.tables where row_number() > 5 ;
Query 20201228_092919_00002_f32n6 failed: java.lang.ClassCastException: class io.prestosql.operator.window.SqlWindowFunction cannot be cast to class io.prestosql.metadata.SqlScalarFunction (io.prestosql.operator.window.SqlWindowFunction and io.prestosql.metadata.SqlScalarFunction are in unnamed module of loader 'app')
java.lang.RuntimeException: java.lang.ClassCastException: class io.prestosql.operator.window.SqlWindowFunction cannot be cast to class io.prestosql.metadata.SqlScalarFunction (io.prestosql.operator.window.SqlWindowFunction and io.prestosql.metadata.SqlScalarFunction are in unnamed module of loader 'app')
	at io.prestosql.metadata.FunctionRegistry.getScalarFunctionInvoker(FunctionRegistry.java:857)
	at io.prestosql.metadata.MetadataManager.getScalarFunctionInvoker(MetadataManager.java:1952)
	at io.prestosql.sql.InterpretedFunctionInvoker.invoke(InterpretedFunctionInvoker.java:63)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitFunctionCall(ExpressionInterpreter.java:1006)
	at io.prestosql.sql.tree.FunctionCall.accept(FunctionCall.java:110)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.evaluateComparisonExpression(ExpressionInterpreter.java:791)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitComparisonExpression(ExpressionInterpreter.java:761)
	at io.prestosql.sql.planner.ExpressionInterpreter$Visitor.visitComparisonExpression(ExpressionInterpreter.java:754)
	at io.prestosql.sql.tree.ComparisonExpression.accept(ComparisonExpression.java:71)
	at io.prestosql.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.prestosql.sql.planner.ExpressionInterpreter.optimize(ExpressionInterpreter.java:294)
	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.rewrite(SimplifyExpressions.java:54)
	at io.prestosql.sql.planner.iterative.rule.SimplifyExpressions.lambda$createRewrite$0(SimplifyExpressions.java:79)
	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$FilterExpressionRewrite.apply(ExpressionRewriteRuleSet.java:220)
	at io.prestosql.sql.planner.iterative.rule.ExpressionRewriteRuleSet$FilterExpressionRewrite.apply(ExpressionRewriteRuleSet.java:201)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.transform(IterativeOptimizer.java:168)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreNode(IterativeOptimizer.java:143)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:108)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreChildren(IterativeOptimizer.java:193)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.exploreGroup(IterativeOptimizer.java:110)
	at io.prestosql.sql.planner.iterative.IterativeOptimizer.optimize(IterativeOptimizer.java:99)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:206)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:195)
	at io.prestosql.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:190)
	at io.prestosql.execution.SqlQueryExecution.doPlanQuery(SqlQueryExecution.java:450)
	at io.prestosql.execution.SqlQueryExecution.planQuery(SqlQueryExecution.java:430)
	at io.prestosql.execution.SqlQueryExecution.start(SqlQueryExecution.java:382)
	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:237)
	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.prestosql.$gen.Presto_unknown____20201228_092751_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.ClassCastException: class io.prestosql.operator.window.SqlWindowFunction cannot be cast to class io.prestosql.metadata.SqlScalarFunction (io.prestosql.operator.window.SqlWindowFunction and io.prestosql.metadata.SqlScalarFunction are in unnamed module of loader 'app')
	at io.prestosql.metadata.FunctionRegistry.specializeScalarFunction(FunctionRegistry.java:864)
	at io.prestosql.metadata.FunctionRegistry.lambda$getScalarFunctionInvoker$3(FunctionRegistry.java:853)
	at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4876)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4871)
	at io.prestosql.metadata.FunctionRegistry.getScalarFunctionInvoker(FunctionRegistry.java:853)
	... 35 more
```

**Comment by omossad** [comment]
Attempt to fix this bug: https://github.com/trinodb/trino/pull/18082

**Comment by omossad** [comment]
pull request to solve this issue (https://github.com/trinodb/trino/pull/18089)

**Comment by omossad** [comment]
Fixed in [#18089](https://github.com/trinodb/trino/pull/18089)
